add unit test for HLT online-DQM plugins

missirol opened the pull request:
#### PR description:

This PR aims to add a unit test involving most of the DQM modules (and services) running in the HLT and providing inputs to the online DQM. The test includes both DQM and harvesting steps, and it only requires that both run without errors (the unit test does not check the outputs).

Small changes are also made in the cpp implementation of a few plugins:
 - the parameter `FolderName` of `PSMonitor` is renamed `folderName` as done for `LumiMonitor` in #39859;
 - default values are added for the parameters of `LumiMonitor` that don't currently have them;
 - a parameter `fillEveryLumiSection` is added to `ThroughputServiceClient` in analogy with `FastTimeServiceClient`.

There are two changes that require feedback (I'm not sure they are correct):

 - I needed to add a call in `ThroughputService` to change the 'scope of the DQM outputs' to `RUN`, otherwise I would not see the `HLT/Throughput` folder in the harvesting output of the unit test; I copied this scope change from `FastTimerService`, and I figure that's what was needed. I'm not sure it is the correct change; clearly the `ThroughputService` outputs were already produced correctly in the online DQM (maybe in that case the default scope is somehow set differently compared to this unit test);
 - I removed the `PSMonitorClient`; it does not produce outputs, it can only issue a warning, and maybe this is not so useful, but also here feedback would be needed.

This PR requires #40325.

Merely technical. No changes expected.

#### PR validation:

Manual tests with new unit test.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

N/A

missirol committed:
removed PSMonitorClient harvester

missirol committed:
set parameter defaults in LumiMonitor

missirol committed:
renamed FolderName to folderName in PSMonitor

missirol committed:
added "fillEveryLumiSection" parameter to ThroughputServiceClient

missirol committed:
added unit test for HLT online-DQM plugins

missirol commented on HLTrigger/Configuration/python/customizeHLTforCMSSW.py:
```diff
@@ -258,5 +269,6 @@ def customizeHLTforCMSSW(process, menuType="GRun"):
  
     process = customizeHLTfor38761(process)
     process = customizeHLTfor40264(process)
+    process = customizeHLTfor40334(process)
```
This customisation is not necessary for the menus in CMSSW. It can become necessary for user menus extracted from `ConfDB`, if those somehow contain a `PSMonitor`.

missirol commented on HLTrigger/Timer/plugins/ThroughputService.cc:
```diff
@@ -85,6 +85,7 @@ void ThroughputService::preGlobalBeginRun(edm::GlobalContext const& gc) {
 
     // define a callback that can book the histograms
     auto bookTransactionCallback = [&, this](DQMStore::IBooker& booker, DQMStore::IGetter&) {
+      auto scope = dqm::reco::DQMStore::IBooker::UseRunScope(booker);
```
This is copied from https://github.com/cms-sw/cmssw/blob/c0d007c5a8e169da2cc6dfea20751bd7215fccb9/HLTrigger/Timer/plugins/FastTimerService.cc#L921

Without this call, the harvesting output of the unit test does not contain a `Throughtput` folder, and I'm not really sure why that is.

missirol replied:
Short version: I think this change is okay. More info below.

I had a look at https://github.com/cms-sw/cmssw/blob/master/DQMServices/Core/README.md, but that didn't really clarify things for me.

If I understand the interface, the DQM scope in `ThroughputService` before the call to
```cpp
auto scope = dqm::reco::DQMStore::IBooker::UseRunScope(booker);
```
corresponds to `scope.oldscope`. The latter returns `1`, which corresponds to `JOB` (and after the call, the scope becomes `RUN`). This seems to match the default set [here](https://github.com/cms-sw/cmssw/blob/d867e73c13e9703c736ceee505fbc14272b58068/DQMServices/Core/interface/DQMStore.h#L556), i.e. `JOB`.

The call to use the `RUN` scope in `FastTimerService` was introduced in #28622 by DQM (maybe `ThoughputService` was simply overlooked in that PR). I don't see a reason why `FastTimerService` and `ThroughputService` should differ in this respect.

With this PR, I managed to produce a ROOT output file with the client `hlt_dqm_clientPB-live_cfg.py` reading `.pb` files produced by re-running a recent HLT menu on 2022 data; in that ROOT output file, I see the `HLT/Throughput` folder, and the plots look as expected. This suggests that this PR does not break the workflow to produce these plots online (which somehow was already working).

Based on the above, I would conclude that this change is okay, even though I don't fully understand it; in particular, I don't know why the `HLT/Throughput` plots were already being produced correctly in the online DQM without this PR.

@cms-sw/dqm-l2 , do you have insight on this?

missirol commented on HLTrigger/Timer/plugins/ThroughputService.cc:
```diff
@@ -95,7 +96,7 @@ void ThroughputService::preGlobalBeginRun(edm::GlobalContext const& gc) {
 
     // book MonitorElement's for this run
-    edm::Service<DQMStore>()->meBookerGetter(bookTransactionCallback);
+    edm::Service<dqm::legacy::DQMStore>()->meBookerGetter(bookTransactionCallback);
```
Maybe this change is unimportant; again it follows what is done in the `FastTimerService`
https://github.com/cms-sw/cmssw/blob/c0d007c5a8e169da2cc6dfea20751bd7215fccb9/HLTrigger/Timer/plugins/FastTimerService.cc#L940

In `ThroughputService`, the following is used
https://github.com/cms-sw/cmssw/blob/c0d007c5a8e169da2cc6dfea20751bd7215fccb9/HLTrigger/Timer/plugins/ThroughputService.h#L32

fwyzard replied:
what difference does it make to use `dqm::legacy` instead of `dqm::reco` ?

fwyzard replied:
answering to myself: none, they are `typedef` one to the other.

missirol replied:
Thanks for the info. Since there is no difference, I will remove the extra `dqm::legacy::` from here.

missirol commented:
test parameters:
  - pull_requests = #40325

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40334/33420

- This PR adds an extra 32KB to repository

cmsbuild commented:
A new Pull Request was created by  @missirol (Marino Missiroli) for master.

It involves the following packages:

- DQM/HLTEvF (**dqm, hlt**)
- DQM/Integration (**dqm**)
- HLTrigger/Configuration (**hlt**)
- HLTrigger/Timer (**hlt**)


@Martin-Grunewald, @emanueleusai, @ahmad3213, @cmsbuild, @missirol, @jfernan2, @syuvivida, @pmandrik, @micsucmed, @rvenditti can you please review it and eventually sign? Thanks.
@batinkov, @battibass, @silviodonato, @mtosi, @Martin-Grunewald, @fwyzard, @threus, @francescobrivio this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented:
@fwyzard , it would be very useful to have your feedback on this PR.

missirol commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c89ae2/29647/summary.html
**COMMIT**: cb811252ab42b7a863556efa54f0ed2cb02c948c
**CMSSW**: CMSSW_13_0_X_2022-12-15-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40334/29647/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 14 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3557521
 - DQMHistoTests: Total failures: 163
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3557336
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-15-1100+c89ae2/54692/triggerResults): no differences found

missirol committed:
use RUN scope in ThroughputService

missirol commented:
test parameters:

missirol commented:
please test

Rebased on IB which includes #40325, and updated following the discussion in https://github.com/cms-sw/cmssw/pull/40334#discussion_r1049973554.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40334/33486

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #40334 was updated. @emanueleusai, @ahmad3213, @Martin-Grunewald, @missirol, @syuvivida, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c89ae2/29715/summary.html
**COMMIT**: 9ff78e4fdd941b22d37972ddbd6a622bcbe141bb
**CMSSW**: CMSSW_13_0_X_2022-12-20-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40334/29715/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555723
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2022-12-20-1100+c89ae2/54752/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40334/33542

- This PR adds an extra 12KB to repository

cmsbuild commented:
Pull request #40334 was updated. @Martin-Grunewald, @emanueleusai, @ahmad3213, @cmsbuild, @missirol, @syuvivida, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

missirol commented:
please test

Rebased, and re-testing.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c89ae2/29779/summary.html
**COMMIT**: cc710bc697ab63720d4945864279189070f72f10
**CMSSW**: CMSSW_13_0_X_2023-01-01-0000/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40334/29779/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555723
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-01-0000+c89ae2/54808/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40334/33543

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #40334 was updated. @Martin-Grunewald, @emanueleusai, @ahmad3213, @cmsbuild, @missirol, @syuvivida, @pmandrik, @micsucmed, @rvenditti can you please check and sign again.

missirol commented:
please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c89ae2/29781/summary.html
**COMMIT**: 444109c9f10e9bbc6f16250d4a1995dbdca4489c
**CMSSW**: CMSSW_13_0_X_2023-01-02-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40334/29781/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

The relvals timed out after 4 hours.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3555720
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-02-1100+c89ae2/54809/triggerResults): no differences found

missirol commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c89ae2/29783/summary.html
**COMMIT**: 444109c9f10e9bbc6f16250d4a1995dbdca4489c
**CMSSW**: CMSSW_13_0_X_2023-01-02-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40334/29783/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 32 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3555748
 - DQMHistoTests: Total failures: 1214
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3554512
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 211 log files, 162 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_0_X_2023-01-02-1100+c89ae2/54810/triggerResults): no differences found

missirol commented:
+hlt

  - adds a unit test involving plugins for online-DQM that are part of online HLT menus (and usually not included in offline HLT menus)
  -  to the best of my knowledge, these changes are okay, but https://github.com/cms-sw/cmssw/pull/40334/files#r1053516164 requires feedback from @cms-sw/dqm-l2

missirol commented:
@cms-sw/dqm-l2, could you please review this PR? Thanks!

emanueleusai commented:
Concerning the removal of the PSMonitorClient, I agree with you, It look like the process just printed a warning if plots are not present. That's odd.

Concerning the ThroughputService, your logic is correct, and I believe what is done in the FastTimerService is correct and can be safely copied to ThroughputService, but the fact that it worked already correctly online makes me suspicious. We could in principle test this online in playback to be extra safe, but we would need a backport...
What do you think?

missirol commented:
> Concerning the ThroughputService, your logic is correct, and I believe what is done in the FastTimerService is correct and can be safely copied to ThroughputService, but the fact that it worked already correctly online makes me suspicious. We could in principle test this online in playback to be extra safe, but we would need a backport...
What do you think?

Opening a backport PR to do this test is probably easy to do (for which cycle ? `12_4_X` ?); I assume the idea is just to open the backport PR to do the test, and then it would not be merged.

On the other hand, I don't think the playback test would help checking the change in `ThroughputService`. If I understand what that test does, it takes existing streamer files and runs the online-DQM clients, so it would not exercise the `ThroughputService` (which runs 'inside' the HLT, when the streamer files are created).

To try and convince myself that this change is okay, I produced streamer files by rerunning the HLT step with this PR, and then manually ran one of the HLT online-DQM clients on those streamer files [1]. In this 'online-like' test, one gets the expected histograms in the `Throughput` folder (both with _and_ without the scope change in `ThroughputService`); in the case of the unit test (which uses `DQMIO`, not `.pb` files), the `Throughput` folder only appears using the 'RUN' scope in `ThroughputService` (i.e. the change in this PR). So, things work with this PR, but I'm not fully clear on what this 'dqm scope' does, and if it somehow acts differently for DQMIO output vs `.pb` output (I take the liberty to tag the author of #28622, @schneiml, just in case he can chime in).

If you think the playback test is useful anyway, I'll open the backport PR to the relevant release cycle.

[1] Tested in `CMSSW_13_0_X_2023-01-12-2300`.
```bash
#!/bin/bash

# scram project CMSSW CMSSW_13_0_X_2023-01-12-2300
# cd CMSSW_13_0_X_2023-01-12-2300/src
# eval `scram runtime -sh`
# git cms-merge-topic cms-sw:40334
# scram build

INPUTFILE=root://eoscms.cern.ch//eos/cms/store/group/dpg_trigger/comm_trigger/TriggerStudiesGroup/STORM/RAW/Run2022F_EphemeralHLTPhysics0_run361468/26ce1488-8c46-436b-becd-6b41535dda79.root

HLTMENU=/users/missirol/test/dev/CMSSW_13_0_0/tmp/test01/cmssw40334/HLT/V3

[ -d run361468 ] || (convertToRaw -f 100 -l 100 -r 361468:172 -o . -- "${INPUTFILE}")

if [ ! -f hlt.py ]; then
  tmpfile=$(mktemp)
  hltConfigFromDB --configName "${HLTMENU}" > "${tmpfile}"
  cat <<@EOF >> "${tmpfile}"

process.load('run361468_cff')

process.hltOnlineBeamSpotESProducer.timeThreshold = int(1e6)

from HLTrigger.Configuration.common import producers_by_type
for producer in producers_by_type(process, 'PSMonitor'):
  if hasattr(producer, 'FolderName'):
    if not hasattr(producer, 'folderName'):
      producer.folderName = producer.FolderName
    del producer.FolderName
@EOF

  edmConfigDump "${tmpfile}" > hlt.py
fi

cmsRun hlt.py &> hlt.log

cmsRun DQM/Integration/python/clients/hlt_dqm_clientPB-live_cfg.py \
  runInputDir=. runNumber=361468 runkey=pp_run \
  scanOnce=True datafnPosition=4

# output file: ./upload/DQM_V0001_HLTpb_R000361468.root
```

emanueleusai commented:
@missirol thank you very much for the detailed explanation. I now understand better your private test, and I agree this is sufficient for the PR to be approved.
I do not know well enough the underlying functioning of the "scopes" to answer your question about the behavior of scopes between DQMIO output vs .pb. So any input from the original developers of the infrastructure is welcome, although I believe @schneiml is not with CMS anymore.
Generally speaking, the way I understand it, the "scope" separates MEs that are filled per-lumi, per-run, or per-job. So if you fill your MEs in the endRun the scope should be set to RUN and so forth.

emanueleusai commented:
+1

- private test works as expected, online test not necessary
- spurious differences in DQM comparisons
- understanding of the underlying structure of scopes can continue separately from this PR imho

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
